File: agent/sidecar/spec.ts
import { z } from "zod";

// MCP server: either stdio (command+args) or HTTP (type=http + URL).
const StdioServerConfig = z.object({
  command: z.string(),
  args: z.array(z.string()).optional(),
  env: z.record(z.string(), z.string()).optional(),
});

const HttpServerConfig = z.object({
  type: z.literal("http"),
  url: z.string().url(),
  headers: z.record(z.string(), z.string()).optional(),
});

export const McpServerConfigSchema = z.union([StdioServerConfig, HttpServerConfig]);

// Auth mode: metadata only. The actual secret is delivered via env vars
// (ANTHROPIC_API_KEY or CLAUDE_CODE_OAUTH_TOKEN) set by the Go runner's
// cmd.Env before exec — NEVER through this spec — so plaintext tokens
// can't leak via stdin captures, /proc snapshots, or a stray log of the
// spec object. See internal/agent/sidecar.go::authEnvFor.
const AuthConfigSchema = z.object({
  mode: z.enum(["subscription", "api_key"]),
});

// JobSpec: the JSON document read from stdin.
export const JobSpecSchema = z.object({
  // Identity (forwarded for log correlation; not sent to the SDK)
  runId: z.string().optional().default(""),
  workflowId: z.string().optional().default(""),

  // Prompt
  prompt: z.string().min(1),
  systemPrompt: z.string().optional(),

  // Model parameters
  model: z.string().min(1),
  maxTurns: z.number().int().positive(),
  maxBudgetUsd: z.number().positive(),

  // Tool config
  toolScope: z.enum(["read_only", "read_write"]).default("read_write"),
  allowedTools: z.array(z.string()).default([]),

  // MCP servers
  mcpServers: z.record(z.string(), McpServerConfigSchema).default({}),

  // Auth
  auth: AuthConfigSchema,

  // Transcript file (optional; Go side opens it too — we duplicate for crash safety)
  transcriptPath: z.string().optional(),

  // Resume a prior SDK session
  sessionId: z.string().optional(),
});

export type JobSpec = z.infer<typeof JobSpecSchema>;
export type McpServerConfig = z.infer<typeof McpServerConfigSchema>;
export type AuthConfig = z.infer<typeof AuthConfigSchema>;

File: internal/agent/spec.go
//go:build !lite

// Package agent provides the Go-side runner and types for the Claude Agent SDK
// integration. The breadbox-agent sidecar binary (built from agent/sidecar/)
// is exec'd per run; this package assembles the JobSpec, streams NDJSON events
// from stdout, and persists the agent_runs row.
package agent

// MCPServerConfig describes one MCP server the sidecar should connect to.
// It models BOTH transports the TS sidecar's zod union accepts (see
// agent/sidecar/spec.ts::McpServerConfigSchema):
//
//   - stdio: Command (+ Args/Env). breadbox itself is always present this way,
//     pointing at the local `breadbox mcp` stdio.
//   - http:  Type="http" + URL (+ Headers). Used by custom connectors, e.g. a
//     remote Gmail MCP reached with an Authorization bearer header.
//
// Command is `omitempty` deliberately: an HTTP entry must NOT serialize an
// empty "command", or the TS union would match it against the stdio variant
// (whose command is just z.string(), which "" satisfies) instead of falling
// through to the HTTP variant.
type MCPServerConfig struct {
	// stdio transport
	Command string            `json:"command,omitempty"`
	Args    []string          `json:"args,omitempty"`
	Env     map[string]string `json:"env,omitempty"`

	// http transport
	Type    string            `json:"type,omitempty"` // "http" when set
	URL     string            `json:"url,omitempty"`
	Headers map[string]string `json:"headers,omitempty"`
}

// AuthConfig carries credentials for one run. Mode picks which env var the
// runner sets on the sidecar process; the other is unset before exec to
// avoid the precedence trap where ANTHROPIC_API_KEY silently wins over
// CLAUDE_CODE_OAUTH_TOKEN.
//
// Token is `json:"-"` deliberately: secrets never travel via the sidecar's
// stdin JSON, only as scoped env vars that Sidecar.Run sets on cmd.Env.
// This closes the leak class where any future `slog.Warn("spec", spec)`
// or similar would otherwise put the plaintext token in logs / OTel.
type AuthConfig struct {
	Mode  string `json:"mode"`  // "subscription" | "api_key" — metadata, safe on the wire
	Token string `json:"-"`     // sk-ant-oat01-… or sk-ant-…; flows to sidecar via env, NEVER JSON
}

// String returns a redacted representation safe for logs / %v formatting.
// Without this an inadvertent `fmt.Sprintf("%+v", spec)` in any handler
// would print the plaintext token. Belt-and-suspenders alongside `json:"-"`.
func (a AuthConfig) String() string {
	suffix := ""
	if n := len(a.Token); n >= 4 {
		suffix = a.Token[n-4:]
	}
	return "AuthConfig{Mode=" + a.Mode + ", Token=…" + suffix + "}"
}

// GoString mirrors String for `%#v` formatting (verbose / debug log style).
func (a AuthConfig) GoString() string { return a.String() }

// JobSpec is the JSON document written to the sidecar's stdin.
// camelCase JSON tags match the TypeScript sidecar's zod schema.
type JobSpec struct {
	// Identity (passed through for log correlation)
	RunID             string `json:"runId"`
	WorkflowID string `json:"workflowId"`

	// Prompt
	Prompt       string `json:"prompt"`
	SystemPrompt string `json:"systemPrompt,omitempty"`

	// Model parameters
	Model        string  `json:"model"`
	MaxTurns     int     `json:"maxTurns"`
	MaxBudgetUsd float64 `json:"maxBudgetUsd"`

	// Tool config
	ToolScope    string   `json:"toolScope"`              // "read_only" | "read_write"
	AllowedTools []string `json:"allowedTools,omitempty"` // MCP tool names; empty = SDK default for scope

	// MCP servers the sidecar should mount
	MCPServers map[string]MCPServerConfig `json:"mcpServers"`

	// Auth
	Auth AuthConfig `json:"auth"`

	// Transcript: absolute path; the sidecar mirrors NDJSON events here.
	TranscriptPath string `json:"transcriptPath,omitempty"`

	// SessionID: when non-empty, the SDK resumes the prior session.
	SessionID string `json:"sessionId,omitempty"`
}
